src/clients/Deluge.ts

```diff
@@ -0,0 +1,226 @@
+import { InjectionResult } from "../constants.js";
+import { CrossSeedError } from "../errors.js";
+import { Label, logger } from "../logger.js";
+import { Metafile } from "../parseTorrent.js";
+import { getRuntimeConfig } from "../runtimeConfig.js";
+import { Searchee } from "../searchee.js";
+import { TorrentClient } from "./TorrentClient.js";
+import { extractCredentialsFromUrl } from "../utils.js";
+import fetch, { Headers, Response } from "node-fetch";
+
+interface DelugeResponse {
+	error?: {
+		message?: string;
+		code?: number;
+	};
+	result?: string | boolean;
+}
+
+export default class Deluge implements TorrentClient {
+	private msgId = 0;
+	private delugeCookie = "";
+	private delugeRpcUrl: URL;
+	private delugeLabel = "cross-seed";
+	private isLabelEnabled: boolean;
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		await this.authenticate();
+		this.isLabelEnabled = await this.labelEnabled();
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<void> {
+		const { delugeRpcUrl } = getRuntimeConfig();
+		const { href, password } = extractCredentialsFromUrl(
+			delugeRpcUrl
+		).unwrapOrThrow(
+			new CrossSeedError("delugeRpcUrl must be percent-encoded")
+		);
+		if (!password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeRpcUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const response = await this.call("auth.login", [password]);
+		if (response === null) {
+			throw new CrossSeedError(
+				`failed to establish a connection to deluge: ${href}`
+			);
+		}
+		if (!response.result) {
+			throw new CrossSeedError(
+				`failed to authenticate with deluge: ${href}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(method: string, params: object, retries = 1) {
+		const { delugeRpcUrl } = getRuntimeConfig();
+		const { href } = extractCredentialsFromUrl(delugeRpcUrl).unwrapOrThrow(
+			new CrossSeedError("delugeRpcUrl must be percent-encoded")
+		);
+		this.msgId = (this.msgId + 1) % 1024;
+		const headers = new Headers({ "Content-Type": "application/json" });
+		if (this.delugeCookie) headers.set("Cookie", this.delugeCookie);
+
+		let response: Response, json: DelugeResponse;
+		try {
+			response = await fetch(href, {
+				body: JSON.stringify({
+					...{ method: method, params: params },
+					id: this.msgId,
+				}),
+				method: "POST",
+				headers: headers,
+			});
+			json = (await response.json()) as DelugeResponse;
+		} catch (networkError) {
+			return null;
+		}
+		if (json?.error?.code === 1 && retries > 0) {
+			await this.authenticate();
+			return this.call(method, params, 0);
+		}
+		if (!response.ok) {
+			return null;
+		}
```

Comment:
I tested a few common scenarios, the JSON-RPC endpoint seems to always return 200 with errors being reported in the actual JSON response.

When the URL is the main webui (not /json), it seems to fail with the "establish connection" error, which would be `response === null`. This could possibly be due to the content-type we use not being acceptable on the main webui, but I'm not sure about this and it's just my assumption.

Reply:
hmm, thinking about it more, if you have Deluge behind nginx or something, maybe nginx could return non-200s. So probably good to have this branch here anyway